Refactor Purity Checker: Preorder Traversal, Visitor Pattern, and Child Linking

HenrikPaetzold opened the pull request.

HenrikPaetzold committed:
Refactor Purity Checker: Preorder Traversal, Visitor Pattern, and Child Linking

    checkValidity() now uses a preorder traversal to collect the entire
    ExpEmbedding tree and filters only relevant node types for purity
    checks

    All ExpEmbedding nodes now expose a children field to support
    generic tree traversal

    Replaced PurityRules with a dedicated visitor implementation
    (ExprPurityVisitor.kt) for handling Assert expressions

jesyspa commented on plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/Purity/ExpVisitor.kt:
Please follow the naming scheme used in the project so far: lowercase for directories, capitals for files.

jesyspa commented on plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/Purity/ExpVisitor.kt:
```diff
@@ -0,0 +1,134 @@
+    fun visitDefault(e: ExpEmbedding): R
+}
+
+internal fun <R> ExpEmbedding.accept(v: ExpVisitor<R>): R = when (this) {
```
Please implement this as a virtual function, so that people know they need to add a new case when they add another class.

jesyspa commented on plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/Purity/ExprPurityVisitor.kt:
```diff
@@ -0,0 +1,106 @@
+    override fun visitInjectionBasedExpEmbedding(e: InjectionBasedExpEmbedding) = e.recurse(this)
+    override fun visitSharingContext(e: SharingContext) = e.recurse(this)
+
+    // TODO: maybe throw errors for nodes that definitely should never appear in an Assert?
```
I think most of these are reasonable to accidentally put in an assert.

jesyspa commented on plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/Purity/ExprPurityVisitor.kt:
```diff
@@ -0,0 +1,106 @@
+    override fun visitIf(e: If) = e.recurse(this)
+    override fun visitElvis(e: Elvis) = e.recurse(this)
+    override fun visitBinaryOperatorExpEmbedding(e: BinaryOperatorExpEmbedding) =
+        e.recurse(this) // TODO: the injection function should be pure per definition?
```
Yes, it is a Viper function.

jesyspa commented on plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/Purity/ExprPurityVisitor.kt:
```diff
@@ -0,0 +1,106 @@
+
+    /* ————— structural nodes without side effects ————— */
+    override fun visitBlock(e: Block) = e.recurse(this)
+    override fun visitIf(e: If) = e.recurse(this)
```
I would mark this as non-pure at this point since we cannot generate pure code for it. (idem for some others)

jesyspa commented on plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/Purity/ExprPurityVisitor.kt:
```diff
@@ -0,0 +1,106 @@
+    override fun visitDefault(e: ExpEmbedding): Boolean = false
+}
+
+private fun ExpEmbedding.recurse(v: ExprPurityVisitor): Boolean =
```
Please pick a more descriptive name.

jesyspa commented on plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/Purity/utils.kt:
```diff
@@ -0,0 +1,32 @@
+import org.jetbrains.kotlin.formver.embeddings.expression.ExpEmbedding
+
+fun ExpEmbedding.preorder(): Sequence<ExpEmbedding> = sequence {
+    val stack = ArrayDeque<Iterator<ExpEmbedding>>()
```
It seems strange that this is necessary.  I would expect the code to go as follows: you yield the current node, then you yield the preorder iterations of all child nodes.  Is there some problem with that?

HenrikPaetzold replied:
No, I just didn't know about yieldAll just until now :)

jesyspa commented on plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:
```diff
@@ -98,9 +99,11 @@ class ProgramConverter(val session: FirSession, override val config: PluginConfi
             body = stmtCtx.convertMethodWithBody(declaration, signature, returnTarget)
         }
+
+        val validity = userFunc.body?.debugExpEmbedding!!.checkValidity()
```
1. Please don't use `debugExpEmbedding` for non-debugging purposes.
2. Make sure to use `validity` somehow.

jesyspa commented on plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/ExpEmbedding.kt:
```diff
@@ -80,6 +80,13 @@ sealed interface ExpEmbedding : DebugPrintable {
+    /**
+     * Responsible for checking the validity of the node
+     */
+    fun checkOwnValidity(): Boolean = true
```
`check` suggests a side effect, I would recommend `isValid`.

jesyspa reviewed (changes requested):
Please make sure to add tests for your new code (in particular, a test that confirms that when we verify an expression that isn't pure, this is caught).

HenrikPaetzold committed:
Refactored preorder to use yieldAll instead of a stack

accept is now a virtual function in each node rather than an extension

checkValidity is now called isValid

HenrikPaetzold committed:
Purity diagnostics reporting:
Functions annotated with @DumpPurityDiagnostics now emit diagnostic information to the output files. This is currently implemented using context parameters, but it can easily be adapted to use regular function parameters if needed.

HenrikPaetzold committed:
renamed Purity->purity
